packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,384 @@
+        # Msg consists of all of the responses for each REST req in JSON format.
+        responses = {}
+        for req_type, req_value in path_value.items():
+            responses[req_type] = req_value["responses"]
+        msg_bytes = json.dumps(responses).encode()
+
+        return Envelope(
+            to=to,
+            sender=sender,
+            protocol_id=protocol_id,
+            message=msg_bytes,
+            context=context,
+        )
+
+    def _process_request(
+        self, http_method: str, url: str, param: str, body: str = "{}"
+    ):
+        """Process incoming API request by packaging into Envelope and sending it in-queue.
+
+        """
+        self._client_id = uuid4().hex
+        protocol_id = PublicId.from_str("fetchai/http:0.1.0")
+        uri = URI(f"http://{self.host}:{self.port}{url}")
+        context = EnvelopeContext(connection_id=self.connection_id, uri=uri)
+        # Prepare the Envelope's message body and encode it into bytes.
+        msg = {
+            "performative": http_method,
+            "path": url,
+            "params": param,
+            "payload": body,
+        }
+        msg_bytes = json.dumps(msg).encode()
+        # Prepare the Envelope itself using the provided variables.
+        envelope = Envelope(
+            to=self.address,
+            sender=self._client_id,
+            protocol_id=protocol_id,
+            context=context,
+            message=msg_bytes,
+        )
+        # Send the Envelope to the Agent's InBox.
+        assert self.in_queue is not None, "Input queue not initialized."
+        assert self.loop is not None, "Loop not initialized."
+        asyncio.run_coroutine_threadsafe(self.in_queue.put(envelope), self.loop)
+
+    def _send_response(self):
+        """Pass the response from out-bounded Envelope back to the front-end client.
+
+        Currently, the response will be written back in the cmd terminal.
+        """
+        # self.wfile.write(self.path.encode())
+        pass
+
+    def connect(self):
+        """
+        Connect.
+
+        Upon HTTP Channel connection, kickstart the HTTP Server.
+        """
+        with self.lock:
+            if self.stopped:
+                self.stopped = False
+                # self.thread.start()
+                # Initializing the internal HTTP server
+                try:
+                    # self.httpd = HTTPServer((self.host, self.port), HTTPHandler(self))
+                    # handler = SimpleHTTPServer.SimpleHTTPRequestHandler
+                    # handler = HTTPHandler.set_channel(self)
+                    # handler.set_channel(self)
+                    self.httpd = HTTPServer(
+                        (self.host, self.port), HTTPHandlerFactory(self)
+                    )
+                    # self.httpd = HTTPServer((self.host, self.port), handler)
+                    logger.info(f"HTTP Server has connected to port {self.port}.")
+                    self.httpd.serve_forever()
```

Comment:
Isn't this blocking here?

Reply:
Not sure what you mean. If you mean by socket, no as long as you pick an available port? I'm currently running on localhost.

Reply:
as in, the call to `connect()` will never return in this setup I believe. So your multiplexer would block here. I think you need to run this in a thread.